guppylang/prelude/_internal/compiler/array.py

```diff
@@ -187,14 +184,54 @@ def build_array_set(
     length: ht.TypeArg,
 ) -> tuple[Wire, Wire]:
     """Builds an array set operation, returning the original element."""
-    # TODO: The `set` operation in hugr returns an either
-    op = UnsupportedOp(
-        op_name="prelude.set",
-        inputs=[array_ty, idx_ty, elem_ty],
-        outputs=[array_ty, elem_ty],
+    sig = ht.FunctionType(
+        [array_ty, ht.USize(), elem_ty],
+        [ht.Sum([[elem_ty, array_ty], [elem_ty, array_ty]])],
+    )
+    if idx_ty != ht.USize():
+        idx = builder.add_op(convert_itousize(), idx)
+    op = ops.ExtOp(
+        hugr.std.PRELUDE.get_op("set"), sig, [length, ht.TypeTypeArg(elem_ty)]
+    )
+    [result] = builder.add_op(op, array, idx, elem)
+    conditional = builder.add_conditional(result)
+    with conditional.add_case(0) as case:
+        error = build_error(case, 1, "array set index out of bounds")
+        case.set_outputs(
+            *build_panic(
+                case, [elem_ty, array_ty], [elem_ty, array_ty], error, *case.inputs()
+            )
+        )
+    with conditional.add_case(1) as case:
+        case.set_outputs(*case.inputs())
+    [elem, array] = conditional
+    return (array, elem)
+
+
+def build_array_get(
+    builder: DfBase[ops.DfParentOp],
+    array: Wire,
+    array_ty: ht.Type,
+    idx: Wire,
+    idx_ty: ht.Type,
+    elem_ty: ht.Type,
+    length: ht.TypeArg,
+) -> Wire:
+    """Builds an array get operation, returning the original element."""
+    sig = ht.FunctionType([array_ty, ht.USize()], [ht.Sum([[], [elem_ty]])])
+    op = ops.ExtOp(
+        hugr.std.PRELUDE.get_op("get"), sig, [length, ht.TypeTypeArg(elem_ty)]
     )
-    array, swapped_elem = iter(builder.add_op(op, array, idx, elem))
-    return array, swapped_elem
+    if idx_ty != ht.USize():
+        idx = builder.add_op(convert_itousize(), idx)
+    [result] = builder.add_op(op, array, idx)
+    conditional = builder.add_conditional(result)
+    with conditional.add_case(0) as case:
+        error = build_error(case, 1, "array get index out of bounds")
+        case.set_outputs(build_panic(case, [], [elem_ty], error))
```

Comment:
AFAICS `set_outputs` takes `*Wire` (i.e. varargs), so the elements need to be individual `Wire`s not a `list[Wire]` - if you had a `list[Wire]` you would have to use `*`. That's in hugr repo.

However, `build_panic` (in guppy repo) claims to return a single `Wire`, so there shouldn't be any `*`ing of that (?!). And the body of `build_panic` returns `builder.add_op(...)`, where `add_op` (in Hugr repo) claims to return a `Node` - so that doesn't look right, either.

So I'm not sure what's going on here!

Reply:
Ah, got it - your latest commit fixes things so they make sense and even `mypy` agrees. (`build_panic` returning `Node` means you can `*` that to get a series of Wires). Looks good now!